free up 3 dispatch keys: make autocast + SparseCSR per-backend keys

bdhirsh opened the pull request:
Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76112

bdhirsh committed:
free up 3 dispatch keys: make autocast + SparseCSR per-backend keys

[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).




[ghstack-poisoned]

albanD commented on c10/core/DispatchKey.h:
```diff
@@ -452,6 +447,27 @@ enum class DispatchKey : uint16_t {
   EndOfSparseBackends = _SparsePrivateUse3,
 
+  // ~~~~~~~~~~~~~~ "SparseCsr" Per-Backend Dispatch keys ~~~~~~~~~~~~~~~~~~~ //
+  // keys starting with an _ are not currently used,
```
Why have these underscores? Why not just have them as proper keys even if no one uses them?
There is a risk that out of tree users might start registering kernels for `_SparseCsrXPU` and then we can never rename it properly?

bdhirsh replied:
Good point on the BC risk.

I originally did it so we could have a reasonable idea of which keys were actually used vs. not used, but you're right - someone might just start using the `_SparseCsrXPU` key out-of-tree without us knowing anyway.

I'll remove all of the leading underscores from the keys as part of this PR too.

albanD commented on c10/core/DispatchKeySet.h:
```diff
@@ -294,7 +294,10 @@ class DispatchKeySet final {
                              DispatchKey::Quantized,
                              DispatchKey::Sparse,
+                             DispatchKey::SparseCsr,
+                             DispatchKey::NestedTensor,
```
NestedTensor?

bdhirsh replied:
Yeah, `NestedTensor` was made to be per-backend a few days ago, and it missed a spot here :(. This one in particular is pretty hard to catch with tests - `has_any(DispatchKeySet)` is really an optimization over calling `has()` multiple times, where it's stated above that you should know what you're doing first - in particular, the keyset that you pass in isn't allowed to have *both* per-backend-functionality bits, and backend bits. This is a debug-only assert that you passed in a valid argument.

I'll add a more explicit comment in `DispatchKey.h` telling you to look for all of the code snippets with `[Note: Per-Backend Functionality Dispatch Keys]` to know where to add new per-backend keys.

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh commented on c10/core/DispatchKey.h:
```diff
@@ -661,37 +709,50 @@ constexpr BackendComponent toBackendComponent(DispatchKey k) {
-        static_cast<uint8_t>(k) -
-        static_cast<uint8_t>(DispatchKey::StartOfDenseBackends));
+        static_cast<uint16_t>(k) -
+        static_cast<uint16_t>(DispatchKey::StartOfDenseBackends));
```
oh this was subtle - a bunch of these casts need to be `uint16_t` now, since the `DispatchKey` enum has more than 64 entries.

I'm not actually sure why this wasn't a problem before - it was caught now because of a `constexpr DispatchKeySet(DispatchKey::AutocastXPU)` call that tripped the compiler.

bdhirsh replied:
Hmm, still looking into the CI errors. I haven't been able to repro them in godbolt

ezyang replied:
I think we should probably start distinguishing between the bitmask (which only has 64 entries) and the extended cross-product set (which has a lot more entries)... maybe it would be better not to list them all out explicitly, and instead have some bitmasking to construct the combined index.

bdhirsh replied:
Can we reasonably do that though without breaking people who are using the existing dispatch key enum vals from the cross product? e.g. `DispatchKey::AutocastXPU` or `DispatchKey::AutogradXLA`

ezyang replied:
Oh yeah, it'd be BC-breaking. But it's just XLA it's not a big deal, we can fix them up.

bdhirsh replied:
What about all of the other backend keys 😃 

(this includes the "dense" keys too, since they're technically not part of the 64-bit range, like `ORT`, `XPU`, `IPU`, etc)

ezyang replied:
Well, another way to make it BC is to grandfather in the old keys (but have a new style that doesn't require writing combinatorially many keys). Still seems better than maintaining huge tables.

ezyang replied:
Returning to this, we should consider refining the types, so we have one type for what can go in the bitset, and another type for the extended set. In any case, I don't think we are relying on the size of the dispatch key in a strict (since they go in bitset) so increasing the size seems fine.

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh committed:
Update on "free up 3 dispatch keys: make autocast + SparseCSR per-backend keys"

Cleans up `DispatchKey.h` so that the remaining dispatch keys that currently have different implementations for different backends (`Autocast` and `SparseCsr`) are properly marked as "per-backend functionality keys".

This also frees up 3 keys worth of space (SparseCsr saves us 1 key, Autocast saves us 2).

Also fixes https://github.com/pytorch/pytorch/issues/76175




[ghstack-poisoned]

bdhirsh commented on test/test_jit_autocast.py:
```diff
@@ -524,10 +524,10 @@ def t_cpu_amp_autocast(x, y):
         self._test_autocast(t_autocast_cuda, "aten::_autocast_to_reduced_precision", x, y)
         self._test_autocast(t_cuda_amp_autocast, "aten::_autocast_to_reduced_precision", x, y)
-        self._test_autocast(t_cpu_amp_autocast, "aten::_autocast_to_reduced_precision", x, y)
+        # self._test_autocast(t_cpu_amp_autocast, "aten::_autocast_to_reduced_precision", x, y)
```
Hey Ed I was going through PR failures (finally trying to land this PR again), and making autocast a per-backend functionality key may be harder than we thought.

Right now, the user API for autocast is something like:
```
with torch.cuda.amp.autocast()
    # only the operations on CUDA tensors will flow through AMP

with torch.cpu.amp.autocast()
    # only the operations on CPU tensors will flow through AMP
```

And there's also `torch.jit.scripting` capture logic to keep this consistent when you script.

If "turning on AMP for a single backend at a time" is really functionality that we want to support, then that won't fit too well into the "per-backend functionality key" model, because we now only have one functionality bit for autocast, that you can add/remove to TLS (`AutocastFunctionality`).

If we don't actually care about supporting that use case though, do you think it's worth trying to change the autocast API (and dealing with the BC breakage?)

If not, I'll just remove the autocast bits from this PR, and land the rest - fixing up the dispatch key printing.
